src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,559 @@
+  // If new guy is already wider than old, no widening
+  if (nt->_widen > ot->_widen) {
+    return nt;
+  }
+
+  if (nt->_widen < Type::WidenMax) {
+    // Returned widened new guy
+    TypeIntPrototype<T, U> prototype{{nt->_lo, nt->_hi}, {nt->_ulo, nt->_uhi}, {nt->_zeros, nt->_ones}};
+    return CT::make(prototype, nt->_widen + 1);
+  }
+
+  // Speed up the convergence by abandoning the bounds, there are only a couple of bits so
+  // they converge fast
+  T min = std::numeric_limits<T>::min();
+  T max = std::numeric_limits<T>::max();
+  U umin = std::numeric_limits<U>::min();
+  U umax = std::numeric_limits<U>::max();
+  U zeros = nt->_zeros;
+  U ones = nt->_ones;
+  if (lt != nullptr) {
+    min = lt->_lo;
+    max = lt->_hi;
+    umin = lt->_ulo;
+    umax = lt->_uhi;
+    zeros |= lt->_zeros;
+    ones |= lt->_ones;
+  }
+  TypeIntPrototype<T, U> prototype{{min, max}, {umin, umax}, {zeros, ones}};
+  return CT::make(prototype, Type::WidenMax);
+}
+template const Type* int_type_widen(const TypeInt* nt, const TypeInt* ot, const TypeInt* lt);
+template const Type* int_type_widen(const TypeLong* nt, const TypeLong* ot, const TypeLong* lt);
+
+// Called by PhiNode::Value during GVN, monotonically narrow the value set, only
+// narrow if the bits change or if the bounds are tightened enough to avoid
+// slow convergence
+template <class CT>
+const Type* int_type_narrow(const CT* nt, const CT* ot) {
+  using T = decltype(CT::_lo);
+  using U = decltype(CT::_ulo);
+
+  if (nt->singleton() || ot == nullptr) {
+    return nt;
+  }
+
+  // If new guy is equal to old guy, no narrowing
+  if (int_type_equal(nt, ot)) {
+    return ot;
+  }
+
+  // If old guy was maximum range, allow the narrowing
+  if (int_type_equal(ot, CT::TYPE_DOMAIN)) {
+    return nt;
+  }
+
+  // Doesn't narrow; pretty weird
+  if (!int_type_subset(ot, nt)) {
+    return nt;
+  }
+
+  // Bits change
+  if (ot->_zeros != nt->_zeros || ot->_ones != nt->_ones) {
+    return nt;
+  }
+
+  // Only narrow if the range shrinks a lot
+  U oc = cardinality_from_bounds(RangeInt<T>{ot->_lo, ot->_hi},
+                                 RangeInt<U>{ot->_ulo, ot->_uhi});
+  U nc = cardinality_from_bounds(RangeInt<T>{nt->_lo, nt->_hi},
+                                 RangeInt<U>{nt->_ulo, nt->_uhi});
+  return (nc > (oc >> 1) + (SMALLINT * 2)) ? ot : nt;
+}
+template const Type* int_type_narrow(const TypeInt* nt, const TypeInt* ot);
+template const Type* int_type_narrow(const TypeLong* nt, const TypeLong* ot);
+
+
+#ifndef PRODUCT
+template <class T>
+static const char* intnamenear(T origin, const char* xname, char* buf, size_t buf_size, T n) {
+  if (n < origin) {
+    if (n <= origin - 10000) {
+      return nullptr;
+    }
+    os::snprintf_checked(buf, buf_size, "%s-" INT32_FORMAT, xname, jint(origin - n));
+  } else if (n > origin) {
+    if (n >= origin + 10000) {
+      return nullptr;
+    }
+    os::snprintf_checked(buf, buf_size, "%s+" INT32_FORMAT, xname, jint(n - origin));
```

Comment:
Was there an explicit choice to use buffers, instead of `outputStream`?

Reply:
I'm not sure, it is just the currently used method.